src/Tactic/Cong.agda

```diff
@@ -0,0 +1,185 @@
+open import Reflection.Name using () renaming (_≟_ to _≟ⁿ_)
+open import Reflection.Pattern hiding (_≟_)
+open import Data.Unit
+open import Data.List
+open import Function.Base
+open import Relation.Nullary
+open import Data.Bool
+open import Data.Product
+open import Data.Maybe hiding (_>>=_)
+
+infixr 2 _≡⌊_⌋_
+
+private
+  extract-f : Term → Term × Bool
+  extract-f term = let (term , state) = ex 0 term false in (vLam "hole" term , state )
+    where
+      open import Category.Monad.State
+      open RawMonadState (StateMonadState Bool)
+
+      ex : ℕ → Term → State Bool Term
+      ex-args : ℕ → List (Arg Term) → State Bool (List (Arg Term))
+      ex-clauses : ℕ → List Clause → State Bool (List Clause)
+
+      ex hole-i (var var-i args) with does (var-i ≥? hole-i)
+      ...                           | true  = do
+                                      args ← ex-args hole-i args
+                                      return $ var (suc var-i) args
+      ...                           | false = do
+                                      args ← ex-args hole-i args
+                                      return $ var var-i args
+      ex hole-i (def f _   )                  with does (f ≟ⁿ quote ⌊_⌋)
+      ex hole-i (def _ (_ ⟅∷⟆ _ ⟅∷⟆ _ ⟨∷⟩ args)) | true  = do
+                                                   put true
+                                                   args ← ex-args hole-i args
+                                                   return $ var hole-i args
+      ex hole-i (def f args)                     | _  = do
+                                                   args ← ex-args hole-i args
+                                                   return $ def f args
+      ex hole-i (con c args) = do
+        args ← ex-args hole-i args
+        return $ con c args
+      ex hole-i (lam v (abs s t)) = do
+        t ← ex (suc hole-i) t
+        return $ lam v $ abs s t
+      ex hole-i (pat-lam cs args) = do
+        cs ← ex-clauses hole-i cs
+        args ← ex-args hole-i args
+        return $ pat-lam cs args
+      ex hole-i (meta x args) = do
+        args ← ex-args hole-i args
+        return $ meta x args
+      ex hole-i (Π[ s ∶ arg v A ] t) = do
+        A ← ex hole-i A
+        t ← ex (suc hole-i) t
+        return $ Π[ s ∶ arg v A ] t
+      ex hole-i (sort s) = return $ sort s
+      ex hole-i (lit l)  = return $ lit l
+      ex hole-i unknown  = return unknown
+
+      ex-args hole-i [] = return []
+      ex-args hole-i (arg v t ∷ as) = do
+        t ← ex hole-i t
+        as ← ex-args hole-i as
+        return $ arg v t ∷ as
+
+      ex-clauses hole-i [] = return []
+      ex-clauses hole-i (clause ps t ∷ cs) = do
+        t ← ex (hole-i + pattern-args-size ps) t
+        cs ← ex-clauses hole-i cs
+        return $ clause ps t ∷ cs
+      ex-clauses hole-i (absurd-clause ps ∷ cs) = do
+        cs ← ex-clauses hole-i cs
+        return $ absurd-clause ps ∷ cs
+
+open import Reflection hiding (map-Args ; _≟_)
```

Comment:
All imports should go at the top of the module.

Reply:
OK. I imported it there so that its definitions of `_>>=_` and friends would not be visible in the above code and cause ambiguities. I'll change it to import at the top then `open` it here instead.